pkg/controller/servicemesh/member/controller_test.go

```diff
@@ -0,0 +1,382 @@
+	member := newMember()
+	member.DeletionTimestamp = &oneMinuteAgo
+	memberRoll := newMemberRoll()
+	memberRoll.Spec.Members = []string{appNamespace}
+
+	cl, _, r := createClientAndReconciler(t, member, memberRoll)
+
+	assertReconcileSucceeds(r, t)
+
+	updatedMemberRoll := test.GetUpdatedObject(cl, memberRoll.ObjectMeta, &maistra.ServiceMeshMemberRoll{}).(*maistra.ServiceMeshMemberRoll)
+	assert.StringArrayEmpty(updatedMemberRoll.Spec.Members, "Expected members list in SMMR to be empty", t)
+
+	updatedMember := test.GetUpdatedObject(cl, member.ObjectMeta, &maistra.ServiceMeshMember{}).(*maistra.ServiceMeshMember)
+	assert.StringArrayEmpty(updatedMember.Finalizers, "Expected finalizers list in SMM to be empty", t)
+}
+
+func TestReconcilePreservesOtherNamespacesInMembersRollWhenAddingMember(t *testing.T) {
+	member := newMember()
+	memberRoll := newMemberRoll()
+	memberRoll.Spec.Members = []string{"other-ns-1", "other-ns-2"}
+
+	cl, _, r := createClientAndReconciler(t, member, memberRoll)
+
+	assertReconcileSucceeds(r, t)
+
+	updatedMemberRoll := test.GetUpdatedObject(cl, memberRoll.ObjectMeta, &maistra.ServiceMeshMemberRoll{}).(*maistra.ServiceMeshMemberRoll)
+	assert.DeepEquals(updatedMemberRoll.Spec.Members, []string{"other-ns-1", "other-ns-2", appNamespace}, "Unexpected members in SMMR", t)
+}
+
+func TestReconcilePreservesOtherNamespacesInMembersRollWhenRemovingMember(t *testing.T) {
+	member := newMember()
+	member.DeletionTimestamp = &oneMinuteAgo
+	memberRoll := newMemberRoll()
+	memberRoll.Spec.Members = []string{"other-ns-1", appNamespace, "other-ns-2"}
+
+	cl, _, r := createClientAndReconciler(t, member, memberRoll)
+
+	assertReconcileSucceeds(r, t)
+
+	updatedMemberRoll := test.GetUpdatedObject(cl, memberRoll.ObjectMeta, &maistra.ServiceMeshMemberRoll{}).(*maistra.ServiceMeshMemberRoll)
+	assert.DeepEquals(updatedMemberRoll.Spec.Members, []string{"other-ns-1", "other-ns-2"}, "Unexpected members in SMMR", t)
+}
+
+func TestReconcileDeletesMemberRollIfItHadCreatedIt(t *testing.T) {
+	member := newMember()
+	member.DeletionTimestamp = &oneMinuteAgo
+	memberRoll := newMemberRoll()
+	memberRoll.Annotations = map[string]string{
+		common.CreatedByKey: controllerName,
+	}
+	memberRoll.Spec.Members = []string{appNamespace}
+
+	cl, _, r := createClientAndReconciler(t, member, memberRoll)
+
+	assertReconcileSucceeds(r, t)
+	test.AssertNotFound(cl, types.NamespacedName{controlPlaneNamespace, common.MemberRollName}, &maistra.ServiceMeshMemberRoll{}, "Expected reconcile to delete the SMMR, but it didn't", t)
+}
+
+func TestReconcileWorksIfMembersRollDoesNotExistWhenRemovingMember(t *testing.T) {
+	member := newMember()
+	member.DeletionTimestamp = &oneMinuteAgo
+
+	_, _, r := createClientAndReconciler(t, member)
+
+	assertReconcileSucceeds(r, t)
+}
+
+// TODO: what if SMMR is created by the controller, but then an additional namespace is added manually.. should the SMMR be deleted when you delete the Member?
```

Comment:
For now, I've added another test that checks if the controller leaves the SMMR alone when someone adds another namespace to it.

The controller already handles such scenarios properly - it leaves the SMMR alone, but deletes it later (after the user removes the manually-added namespace from it). I don't really think we need to complicate things further. Most users will only ever use one of the two options - either create and manage the SMMR manually, or use ServiceMeshMember objects instead of SMMR.

Reply:
Makes sense, I'm happy to keep it as is